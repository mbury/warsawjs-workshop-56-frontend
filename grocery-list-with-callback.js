export function init(view) {
  const mountNode = view || document;

  const store = getGroceryListStore();

  store.addItem('ser');
  store.addItem('chleb');

  const addInput = mountNode.querySelector('.add-input');
  const addButton = mountNode.querySelector('.add-button');
  const groceryList = mountNode.querySelector('.grocery-list');
  const filterOption = mountNode.querySelector('.filter-options');

  addButton.addEventListener('click', addItemHandler);
  document.addEventListener('DOMContentLoaded', initHandler);
  filterOption.addEventListener('change', selectFilterHandle);

  function addItemHandler(e) {
    e.preventDefault();

    store.addItem(addInput.value, (itemData) => {
      const newItemNode = getItemNodeFrom(itemData);
      groceryList.appendChild(newItemNode);
      addInput.value = '';
    });
  }

  function selectFilterHandle(e) {
    const checklist = groceryList.childNodes;
    checklist.forEach(function (item) {
      switch (e.target.value) {
        case 'all':
          item.style.display = 'flex';
          break;
        case 'completed':
          if (item.classList.contains('completed')) {
            item.style.display = 'flex';
          } else {
            item.style.display = 'none';
          }
          break;
        case 'uncompleted':
          if (!item.classList.contains('completed')) {
            item.style.display = 'flex';
          } else {
            item.style.display = 'none';
          }
      }
    });
  }

  function initHandler() {
    const data = store.getState();

    data.groceryList.forEach(function (item) {
      const newItem = getItemNodeFrom(item);

      groceryList.appendChild(newItem);
    });
  }

  function getItemNodeFrom(item) {
    const newItem = document.createElement('li');
    newItem.classList.add('grocery-list-item');

    const text = document.createElement('span');
    text.textContent = item.text;

    text.classList.add('grocery-list-item-text');
    newItem.appendChild(text);

    const completedButton = document.createElement('button');
    completedButton.innerHTML = `<i class="fas fa-check"></i>`;
    completedButton.classList.add('complete-btn');
    completedButton.addEventListener('click', () =>
      store.completeItem(item.id, () => newItem.classList.toggle('completed'))
    );
    newItem.appendChild(completedButton);

    const deleteButton = document.createElement('button');
    deleteButton.innerHTML = `<i class="fas fa-trash"></i>`;
    deleteButton.classList.add('delete-btn');
    deleteButton.addEventListener('click', () =>
      store.deleteItem(item.id, () => newItem.remove())
    );
    newItem.appendChild(deleteButton);
    return newItem;
  }
}

function getGroceryListStore(initState) {
  const store = {};
  store.state = { groceryList: initState || [] };
  store.getState = () => store.state;
  store.addItem = (text, callback) => {
    const newGroceryItem = {
      id: new Date(),
      text: text,
      completed: false,
    };
    const newGroceryList = [...store.state.groceryList, newGroceryItem];
    const nextState = { groceryList: newGroceryList };
    store.state = nextState;

    if (callback) {
      callback(newGroceryItem);
    }
  };

  store.deleteItem = (id, callback) => {
    const groceryList = store.state.groceryList.filter((i) => {
      return i.id !== id;
    });

    const nextState = {
      groceryList,
    };

    store.state = nextState;

    if (callback) {
      callback();
    }
  };

  store.completeItem = (id, callback) => {
    const groceryList = store.state.groceryList.map((i) => {
      if (i.id === id) {
        return { ...i, completed: !i.completed };
      }
      return i;
    });

    const nextState = {
      groceryList,
    };
    store.state = nextState;

    if (callback) {
      callback();
    }
  };

  return store;
}
