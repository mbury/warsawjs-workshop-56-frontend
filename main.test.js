import {
  getByText,
  getByPlaceholderText,
  getByRole,
  fireEvent,
} from '@testing-library/dom';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';

import { init } from './grocery-list-with-callback';

function getDOMTemplate() {
  const div = document.createElement('div');
  div.innerHTML = `
    <form>
      <input placeholder="dodaj do listy" type="text" class="add-input" />
      <button class="add-button" type="submit">
        <i class="fas fa-plus"></i>
      </button>
      <div class="select">
        <select name="filters" class="filter-options">
          <option value="all">Wszystkie</option>
          <option value="completed">Kupione</option>
          <option value="uncompleted">Niekupione</option>
        </select>
      </div>
    </form>
    <section id="grocery-list-root">
      <ul class="grocery-list"></ul>
    </section>
  `;
  init(div);

  return div;
}

test('should add new item to list', async () => {
  const item = 'chleb';
  const container = getDOMTemplate();
  const input = getByPlaceholderText(container, 'dodaj do listy');

  fireEvent.change(input, { target: { value: item } });

  getByRole(container, 'button').click();
  expect(getByText(container, item)).toHaveTextContent(item);
  expect(container).toMatchSnapshot();
});

test('should hide uncompleted item after filter select', async () => {
  const item = 'chleb';
  const container = getDOMTemplate();
  const input = getByPlaceholderText(container, 'dodaj do listy');
  fireEvent.change(input, { target: { value: item } });
  getByRole(container, 'button').click();

  userEvent.selectOptions(getByRole(container, 'combobox'), 'completed');

  expect(getByText(container, item)).toHaveTextContent(item);
  expect(container).toMatchSnapshot();
});
