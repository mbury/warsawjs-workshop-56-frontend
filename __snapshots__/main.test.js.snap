// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add new item to list 1`] = `
<div>
  
    
  <form>
    
      
    <input
      class="add-input"
      placeholder="dodaj do listy"
      type="text"
    />
    
      
    <button
      class="add-button"
      type="submit"
    >
      
        
      <i
        class="fas fa-plus"
      />
      
      
    </button>
    
      
    <div
      class="select"
    >
      
        
      <select
        class="filter-options"
        name="filters"
      >
        
          
        <option
          value="all"
        >
          Wszystkie
        </option>
        
          
        <option
          value="completed"
        >
          Kupione
        </option>
        
          
        <option
          value="uncompleted"
        >
          Niekupione
        </option>
        
        
      </select>
      
      
    </div>
    
    
  </form>
  
    
  <section
    id="grocery-list-root"
  >
    
      
    <ul
      class="grocery-list"
    >
      <li
        class="grocery-list-item"
      >
        <span
          class="grocery-list-item-text"
        >
          chleb
        </span>
        <button
          class="complete-btn"
        >
          <i
            class="fas fa-check"
          />
        </button>
        <button
          class="delete-btn"
        >
          <i
            class="fas fa-trash"
          />
        </button>
      </li>
    </ul>
    
    
  </section>
  
  
</div>
`;

exports[`should hide uncompleted item after filter select 1`] = `
<div>
  
    
  <form>
    
      
    <input
      class="add-input"
      placeholder="dodaj do listy"
      type="text"
    />
    
      
    <button
      class="add-button"
      type="submit"
    >
      
        
      <i
        class="fas fa-plus"
      />
      
      
    </button>
    
      
    <div
      class="select"
    >
      
        
      <select
        class="filter-options"
        name="filters"
      >
        
          
        <option
          value="all"
        >
          Wszystkie
        </option>
        
          
        <option
          value="completed"
        >
          Kupione
        </option>
        
          
        <option
          value="uncompleted"
        >
          Niekupione
        </option>
        
        
      </select>
      
      
    </div>
    
    
  </form>
  
    
  <section
    id="grocery-list-root"
  >
    
      
    <ul
      class="grocery-list"
    >
      <li
        class="grocery-list-item"
        style="display: none;"
      >
        <span
          class="grocery-list-item-text"
        >
          chleb
        </span>
        <button
          class="complete-btn"
        >
          <i
            class="fas fa-check"
          />
        </button>
        <button
          class="delete-btn"
        >
          <i
            class="fas fa-trash"
          />
        </button>
      </li>
    </ul>
    
    
  </section>
  
  
</div>
`;
